File: modules/zivid/_calibration/_multi_camera.py
import _zivid


class MultiCameraResidual:
    def __init__(self, internal_multi_camera_residual):
        if not isinstance(
            internal_multi_camera_residual, _zivid.calibration.MultiCameraResidual
        ):
            raise TypeError(
                "Unsupported type: {recieved_type}".format(
                    recieved_type=type(internal_multi_camera_residual)
                )
            )
        self.__impl = internal_multi_camera_residual

    def translation(self):
        return self.__impl.translation()

    def __str__(self):
        return str(self.__impl)


class MultiCameraOutput:
    def __init__(self, internal_multi_camera_output):
        if not isinstance(
            internal_multi_camera_output, _zivid.calibration.MultiCameraOutput
        ):
            raise TypeError(
                "Unsupported type: {recieved_type}".format(
                    recieved_type=type(internal_multi_camera_output)
                )
            )
        self.__impl = internal_multi_camera_output

    def valid(self):
        return self.__impl.valid()

    def __bool__(self):
        return bool(self.__impl)

    def transforms(self):
        return self.__impl.transforms()

    def residuals(self):
        return self.__impl.residuals()

    def __str__(self):
        return str(self.__impl)


def calibrate_multi_camera(detection_results):
    return MultiCameraOutput(
        _zivid.calibration.calibrate_multi_camera(
            [
                detection_result._DetectionResult__impl
                for detection_result in detection_results
            ]
        )
    )

File: src/Calibration/MultiCamera.cpp
#include <ZividPython/Calibration/MultiCamera.h>
#include <ZividPython/Matrix.h>

#include <pybind11/pybind11.h>
#include <pybind11/stl.h>

namespace py = pybind11;

namespace ZividPython
{
    void wrapClass(pybind11::class_<Zivid::Calibration::MultiCameraResidual> pyClass)
    {
        pyClass.def("translation", &Zivid::Calibration::MultiCameraResidual::translation);
    }
    void wrapClass(pybind11::class_<Zivid::Calibration::MultiCameraOutput> pyClass)
    {
        pyClass.def("valid", &Zivid::Calibration::MultiCameraOutput::valid)
            .def("transforms",
                 [](const Zivid::Calibration::MultiCameraOutput &calibrationOutputs) {
                     auto converted_transforms = std::vector<Eigen::Matrix<float, 4, 4, Eigen::RowMajor>>();
                     for(const auto &calibrationOutput : calibrationOutputs.transforms())
                     {
                         converted_transforms.emplace_back(Conversion::toPy(calibrationOutput));
                     }
                     return converted_transforms;
                 })
            .def("residuals", &Zivid::Calibration::MultiCameraOutput::residuals);
    }
} // namespace ZividPython
